### DISCUSSION TITLE
HW Accelerated encoding

### DISCUSSION THREAD

**Discussion by pptudela** [mediacms-io/mediacms#314] [Q&A]
Has anyone successfully configured ffmpeg with Quick Sync or Nvenc to use hardware acceleration for encoding the different quality copies?

I presume it's just a matter changing the ffmpeg commands on tasks.py and adequately compiling ffmpeg, but was wondering if someone has done this before.

**Comment by nassaba-ai** [comment]
Was hoping to find an answer to that, and just find your question. It's not only matter of `tasks.py` but `helpers.py` too.

I'm trying some combinations. Without luck. Here's a sample of the logs:

`logs/debug.log:FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg -hwaccel qsv'`

It will need more effort ...
![bitmoji](https://sdk.bitmoji.com/render/panel/38e3bc7c-223a-4cd4-9c00-b7a942bcc2d0-2fd43d40-3564-4a8f-8c54-3e0276b838e4-v1.png?transparent=1&palette=1&width=246)

**Comment by Anduin2017** [comment]
Hi, any updates?

I bought an Nvidia Tesla P40. However, seems I couldn't benefit from it for running MediaCMS.

Any solution to use hardware accelerated encoding?

**Reply by mgogoulos** [reply]
Have no experience with what it is involved but believe it has to be a setting to set ffmpeg know about using GPU.

**Reply by Anduin2017** [reply]
To add support for GPU acceleration in MediaCMS, the first step is to identify which GPU acceleration solutions are supported by the system it is running on. This varies depending on the hardware and software configuration of the system and can include Nvidia NVENC, Intel QSV, AMD AMF, and Apple videotoolbox.

Once the available GPU acceleration options are identified, a new setting can be added to the encoder settings menu in MediaCMS. This setting will allow users to choose which GPU acceleration solution they want to use for encoding. This setting can be implemented as a dropdown with the available GPU acceleration options listed.

When a user selects a GPU acceleration option, the encoder will pass the corresponding codec to ffmpeg to use for encoding. For example, if a user selects Nvidia NVENC, the encoder will pass the nvenc codec to ffmpeg.

It's important to note that not all GPU acceleration options will be available on every system. If a system does not support GPU acceleration, the encoder can default to using the CPU for encoding instead.

Overall, adding support for GPU acceleration in MediaCMS requires adding a new setting to the encoder, identifying the available GPU acceleration options, and passing the selected codec to ffmpeg for encoding.

**Reply by Anduin2017** [reply]
For example, to encode a video to HEVC with CPU, we can run:

```bash
ffmpeg -i input.mp4 -preset slow -codec:a copy -codec:v libx265 -crf 20
```

But with GPU, we need to run:

```bash
ffmpeg -i input.mp4 -preset slow -codec:a copy -codec:v hevc_nvenc -rc:v vbr -cq:v 20 -rc-lookahead 10 -profile:v main10
```

![image](https://user-images.githubusercontent.com/19531547/234033050-6523d69b-270c-4702-8de7-48e518e79f3c.png)


According to my test, for CPU E5-2696v4 16 cores and Nvidia Tesla P40, using GPU, P40 has 5 times performance advantage.

**Reply by mgogoulos** [reply]
Hi @Anduin2017 , thanks for sharing this interesting info! 

One would have to edit files/helpers.py and add the parameters manually for now to ffmpeg, since this is not possible through a web interface.
Cheers

**Reply by MyKeyTeeNZ** [reply]
Hi @mgogoulos - thanks for that, do you have an example of what we need to change in the helpers.py file for this to work?  My GPU is Nvidia, just not sure where to edit at this point?  Thanks in advance!

**Reply by mgogoulos** [reply]
have a look on files.helpers.produce_ffmpeg_commands, this is calling files.helpers.get_base_ffmpeg_commands , and is providing the ffmpeg command that is run for a single encoding (or commands, because there is a two pass encoding in some cases)

there's a lot of logic there and it's not easy to decouple, however it's a start! if you know what extra arguments ffmpeg is going to need, or need to adjust existing ones, that's the place to look

**Reply by dmurphynj** [reply]
[helpers-vaapi.txt](https://github.com/mediacms-io/mediacms/files/14364087/helpers-vaapi.txt)

Here is an updated helpers.py that I hacked FFmpeg vaapi support into.

I'm not a coder, don't play one on TV, nor am I an ffmpeg expert, but seems to be working ... haven't tested it much but it's a starting point.

Run a diff against the existing helpers.py and you can see what was changed.

**Comment by BadVolf** [comment]
Hey guys, here you go:

Rename it to helpers.py
[helpers-hvenc.txt](https://github.com/mediacms-io/mediacms/files/15080732/helpers-hvenc.txt)

This is working with my the NVidia RTX 3070IT Super. you will see it tests if the media should be hardware encoded and if so, it has filters for h.264 and h265

`    if use_hardware_encoding:
        encoder = "h264_nvenc" if codec == "h264" else "hevc_nvenc"`
        
Sorry, I don't know anything about github, just thought I would share this with you.

Let me know if you have any improvements.

**Reply by IllegalD** [reply]
> Hey guys, here you go:
> 
> Rename it to helpers.py [helpers-hvenc.txt](https://github.com/mediacms-io/mediacms/files/15080732/helpers-hvenc.txt)
> 
> This is working with my the NVidia RTX 3070IT Super. you will see it tests if the media should be hardware encoded and if so, it has filters for h.264 and h265
> 
> ` if use_hardware_encoding: encoder = "h264_nvenc" if codec == "h264" else "hevc_nvenc"`
> 
> Sorry, I don't know anything about github, just thought I would share this with you.
> 
> Let me know if you have any improvements.

Your two configs have been a godsend, thanks mate! For anyone else trying to get hardware transcoding working, I found it easiest to replace ffmpeg with jellyfin-ffmpeg. Have had it working great on vaapi and nvenc using these two configs. It also helps to bump up the bitrates too, as software transcoding still looks a little better.

**Reply by dmurphynj** [reply]
> Your two configs have been a godsend, thanks mate! For anyone else trying to get hardware transcoding working, I found it easiest to replace ffmpeg with jellyfin-ffmpeg. Have had it working great on vaapi and nvenc using these two configs. It also helps to bump up the bitrates too, as software transcoding still looks a little better.

Glad it helps!  I agree - using the jellyfin-ffmpeg package is a big, big help since it has everything compiled in.

I'm doing some interesting tricks here - my mediacms instance runs in a VM, under proxmox, on an Intel N100 CPU.  It took quite a bit of work to get GPU passthrough to work, but I'm passing through the internal GPU in the N100 chip to the mediacms VM, so ffmpeg has direct access to it.

Complex config but working great - huge, huge improvement in transcode time vs. software-only.

The N100 CPU is ridiculously impressive for the cost...  There's certainly better CPUs (and GPUs) out there, but for a 6-watt CPU?  Insanely good.

**Reply by IllegalD** [reply]
I'm running it in an Ubuntu LXC an old Nuc with an i3-6100U, the vaapi performance is fantastic compared to CPU transcoding!

**Reply by mabruzzo3169** [reply]
Anyone have any interest in helping a new user replace ffmpeg with jellyfin-ffmpeg?

**Reply by stp14** [reply]
I am clearly missing something here. I replaced ffmpeg with jellyfin-ffmpeg and am using @BadVolf's helpers.py file - but I'm getting an error in celery_long.log that `name 'use_hardware_encoding' is not defined` no matter how many times and different locations I try to manually set it to `True`.  I've tried to set it globally, I've tried setting it inside the various functions, I've tried passing it as an argument to the functions... So far, no luck.

I just get a bunch of these:
```
[2024-12-20 15:29:23,829: ERROR/ForkPoolWorker-6] Task encode_media[6d7d8851-08fc-44e9-86da-f3eb3f348176] raised unexpected: NameError("name 'use_hardware_encoding' is not defined")
Traceback (most recent call last):
  File "/home/mediacms.io/lib/python3.11/site-packages/celery/app/trace.py", line 477, in trace_task
    R = retval = fun(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^
  File "/home/mediacms.io/lib/python3.11/site-packages/celery/app/trace.py", line 760, in __protected_call__
    return self.run(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mediacms.io/mediacms/files/tasks.py", line 274, in encode_media
    tfpass = create_temp_file(suffix=".{0}".format(profile.extension), dir=temp_dir)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mediacms.io/mediacms/files/helpers.py", line 749, in produce_ffmpeg_commands
  File "/home/mediacms.io/mediacms/files/helpers.py", line 550, in get_base_ffmpeg_command
NameError: name 'use_hardware_encoding' is not defined
```

EDIT: Update, I'm an idiot. I was editing files on the host system and not restarting the container. But basically I'm just hardcoding `use_hardware_encoding = True` because I don't see anywhere else that that value is getting set. Am I just missing something obvious?

After manually setting that, now I'm having a new error, where it seems my container is missing `libnvcuvid.so` and `libnvidia-encode.so`. Not sure how/why though - they're on my host and the container can clearly see the GPU.

EDIT 2: Still not sure why those libraries were missing... but I did some shenanigans to get around that (copied them into the container). Still no joy on a lot of video files. I thought it was my GPU not supporting 10-bit video but that's not the case. Maybe something about the pixel format, but I'm not 100% sure there either. I feel like I'm approaching the point where I need to take a step back and redeploy these containers.

**Reply by HeneryH** [reply]
> > Your two configs have been a godsend, thanks mate! For anyone else trying to get hardware transcoding working, I found it easiest to replace ffmpeg with jellyfin-ffmpeg. Have had it working great on vaapi and nvenc using these two configs. It also helps to bump up the bitrates too, as software transcoding still looks a little better.
> 
> Glad it helps! I agree - using the jellyfin-ffmpeg package is a big, big help since it has everything compiled in.
> 
> I'm doing some interesting tricks here - my mediacms instance runs in a VM, under proxmox, on an Intel N100 CPU. It took quite a bit of work to get GPU passthrough to work, but I'm passing through the internal GPU in the N100 chip to the mediacms VM, so ffmpeg has direct access to it.
> 
> Complex config but working great - huge, huge improvement in transcode time vs. software-only.
> 
> The N100 CPU is ridiculously impressive for the cost... There's certainly better CPUs (and GPUs) out there, but for a 6-watt CPU? Insanely good.

@dmurphynj Dennis, I am trying to do the same and found the passthrough has been a challenge.  Have you created any notes that describe your solution?  I saw an interesting blog about passing through the GPU in what seems like an easy alternate method compared to the full GPU passthrough.  I am in the middle of trying this [Proxmox: Install Plex LXC with HW Transcoding/HDR Tone Mapping](https://www.derekseaman.com/2023/04/proxmox-plex-lxc-with-alder-lake-transcoding.html)

**Reply by dmurphynj** [reply]
If you follow this to a T, it's what I'm using to make it work.  Hasn't done wrong by me yet -- there's a LOT of different versions of this floating around, but Derek's Win11 version has been the most comprehensive and accurate thus far IMO.

I should also note that I'm not running any containers - this is a straight up VM.

https://www.derekseaman.com/2024/07/proxmox-ve-8-2-windows-11-vgpu-vt-d-passthrough-with-intel-alder-lake.html


edit: Added to say, of course you have to sub out Win11 for whichever Linux flavor you have -- and then install the DKMS i915 module inside that as well.  I'm using Ubuntu, so the commands in my VM are something like:

```
sudo bash
cd ~
git clone https://github.com/strongtz/i915-sriov-dkms.git
apt install build-* dkms
cd ~/i915-sriov-dkms
dkms add .

dkms install -m i915-sriov-dkms -v $(cat VERSION) --force
dkms status

cp -a /etc/default/grub{,.bak}
sed -i '/^GRUB_CMDLINE_LINUX_DEFAULT/c\GRUB_CMDLINE_LINUX_DEFAULT="quiet i915.enable_guc=3"' /etc/default/grub
update-grub
update-initramfs -u -k all

```

**Reply by HeneryH** [reply]
Eh, I am running an 8th gen or so CPU.  That method I think requires a 12th gen or newer.

**Reply by HeneryH** [reply]
I am using one of his other blogs about using the plex helper script that shows that a GPU can be passed in automatically.  But I am not 100% sure of it.  [Proxmox: Install Plex LXC with HW Transcoding/HDR Tone Mapping](https://www.derekseaman.com/2023/04/proxmox-plex-lxc-with-alder-lake-transcoding.html)

> In this post I cover how to setup a Plex Media Server (PMS) LXC on Proxmox with hardware transcoding and HDR tone mapping. Contrary to popular belief, the Plex LXC does not need to be configured with PCIe passthrough for the GPU. So NO Proxmox IOMMU host configuration is needed to allow the Plex LXC to do hardware transcoding and HDR tone mapping. This procedure uses the awesome [tteck Promox Plex LXC script](https://tteck.github.io/Proxmox/) to mostly automate the setup of a Plex LXC with the needed GPU drivers and LXC configuration to enable these GPU offload features.

**Reply by dmurphynj** [reply]
> Eh, I am running an 8th gen or so CPU. That method I think requires a 12th gen or newer.

Frankly, the N100 based mini PCs are so cheap and effective, I'd consider replacing that 8th gen CPU with it.  Shockingly good - I ditched an AMD 5950X rack mount server with a Beelink N100 minipc and the power savings alone are pretty close to paying for it. :-)

**Reply by feasly** [reply]
> Hey guys, here you go:
> 
> Rename it to helpers.py [helpers-hvenc.txt](https://github.com/mediacms-io/mediacms/files/15080732/helpers-hvenc.txt)
> 
> This is working with my the NVidia RTX 3070IT Super. you will see it tests if the media should be hardware encoded and if so, it has filters for h.264 and h265
> 
> ` if use_hardware_encoding: encoder = "h264_nvenc" if codec == "h264" else "hevc_nvenc"`
> 
> Sorry, I don't know anything about github, just thought I would share this with you.
> 
> Let me know if you have any improvements.

Many thanks! Is there a newer version? I got a 502 bad gateway if I use your file for my helpers.py, I tried to edit the differences of the stock helpers.py manually, but still MediaCMS only uses the CPU.

**Comment by Bewinxed** [comment]
How to make this work in the docker-compose image?

**Comment by mrjoshuap** [comment]
I've successfully added hardware encoding for Rockchip RK3588S arm64's, which is similar in modifications.  I was however interested in thoughts on how to abstract the encoding profiles to make it a little easier to support different hardware encoders.

For my purposes, I had a custom build version of ffmpeg, bento4, and pass through transcoding hardware into the worker containers.  If you're using a container deployment, make sure you pass through the devices into your docker-compose file.  For instance, on RK3588s, you have to add the following devices to celery_worker:

```
  celery_worker:
    devices:
      - "/dev/dri:/dev/dri:rwm"
      - "/dev/dma_heap:/dev/dma_heap:rwm"
      - "/dev/rga:/dev/rga:rwm"
      - "/dev/mpp_service:/dev/mpp_service:rwm"
      - "/dev/video-dec0:/dev/video-dec0:rwm"
      - "/dev/video-enc0:/dev/video-enc0:rwm"

```

If you're using Nvidia or other hardware encoders, the devices will be different but will probably still be required...

It's also worth mentioning that when you are testing encoding profiles, grabbing the exact ffmpeg command being run is extremely important because a permissions problem on your devices usually ends up with a Segmentation Fault and basically no usable error messages in any of the logs.

**Comment by HeneryH** [comment]
Sorry if this is old or stale.  Trying to get my intel embedded quicksync gpu that is being passed through Proxmox to work.

I see this in my helpers.py:

```
    if codec == "h264":
        encoder = "libx264"
        # ext = "mp4"
    elif codec in ["h265", "hevc"]:
        encoder = "libx265"
        # ext = "mp4"
    elif codec == "vp9":
        encoder = "libvpx-vp9"
        # ext = "webm"
    else:
        return False
 
   cmds = []
    for pass_number in passes:
        cmds.append(
            get_base_ffmpeg_command(
                media_file,
                output_file=output_filename,
                has_audio=media_info.get("has_audio"),
                codec=codec,
                encoder=encoder,
                audio_encoder=AUDIO_ENCODERS[codec],
                target_fps=target_fps,
                interlaced=interlaced,
                target_height=resolution,
                target_rate=target_rate,
                target_rate_audio=AUDIO_BITRATES[codec],
                pass_file=pass_file,
                pass_number=pass_number,
                enc_type=enc_type,
                chunk=chunk,
            )
        )
    return cmds
```

And I see this from my google search on how to do it...

### FFmpeg Intel Acceleration
To enable Intel hardware acceleration for FFmpeg in MediaCMS, you need to configure FFmpeg to use the Intel Quick Sync Video (QSV) hardware acceleration. Here are the steps and considerations:
- Intel Quick Sync Video (QSV): Intel's hardware acceleration technology for video encoding and decoding. To use QSV with FFmpeg, you need to compile FFmpeg with support for libmfx (Intel Media SDK). Ensure your system has the necessary drivers and libraries installed.
- MediaCMS Configuration: For MediaCMS, you will need to modify the helpers.py file to include the QSV parameters manually since the web interface does not support this configuration. This involves adding the appropriate FFmpeg command parameters for QSV.
- Example FFmpeg Command: An example command for encoding a video using QSV would be:
`ffmpeg -i input.mp4 -c:v h264_qsv -preset slow -c:a copy output.mp4`
`ffmpeg -i input.mp4 -c:v h265_qsv -preset slow -c:a copy -global_quality 20 -b:v 0 output.mp4`
This command uses the h264_qsv codec for video encoding with the slow preset to ensure good quality.
- Performance Considerations: Intel QSV can significantly improve encoding and decoding speed compared to software-only processing. However, the performance gain depends on the specific Intel CPU and its support for QSV.
- Testing and Debugging: After making changes, test the configuration thoroughly to ensure that hardware acceleration is working correctly. Check logs for any errors or warnings related to FFmpeg or hardware acceleration.

Trying to figure out the proper way to inject the 
`c:v h264_qsv -preset slow`
or
`c:v h265_qsv -preset slow`

**Reply by HeneryH** [reply]
Hey @dmurphynj , I am looking at your diff for the helpers.py file.  Can you tell me what the intent of this change was?  Not sure if I might be comparing and older version. 

Comparing files helpers.py and HELPERS-VAAPI.TXT
***** helpers.py
        stdout = run_command(cmd).get("out")
        stream_size = sum([int(line.replace("|", "")) for line in stdout.split("\n") if line != ""])
        video_bitrate = round((stream_size * 8 / 1024.0) / video_duration, 2)
***** HELPERS-VAAPI.TXT
        stdout = run_command(cmd).get("out")
        stream_size = sum([int(line) for line in stdout.split("\n") if line != ""])
        video_bitrate = round((stream_size * 8 / 1024.0) / video_duration, 2)
*****

**Reply by dmurphynj** [reply]
That looks like two different versions of helpers.py possibly?

I'm looking at the original helpers.py I spawned the VAAPI version from, and I didn't modify that code.

It was as follows:
```
        stdout = run_command(cmd).get("out")
        stream_size = sum([int(line) for line in stdout.split("\n") if line != ""])
        video_bitrate = round((stream_size * 8 / 1024.0) / video_duration, 2)
```

Not sure when it changed, but must be after the version I'm using.

**Reply by HeneryH** [reply]
Thank you.  I figured.  All the other diffs made sense to me.

Have you found a way to check if the GPU is running on transcode?  I tried using intel-gpu-top on the proxmox host but It never shows any activity.   And the CPU goes to 200% cpu with ffmpeg.

**Reply by dmurphynj** [reply]
Yeah, I use 'vainfo' to make sure all the profiles are available ...

And then I double check ffmpeg to make sure it's actually using the hardware .... something like this:

```
 time /usr/lib/jellyfin-ffmpeg/ffmpeg -c:v h264 -I source.mp4 test_noaccel.mp4
 time /usr/lib/jellyfin-ffmpeg/ffmpeg -hwaccel qsv -c:v h264_qsv -i source.mp4 -vf 'vpp_qsv=framerate=60,scale_qsv=w=1920:h=108' -c:v h264_qsv test_accel.mp4
```

Very primitive but this is what happens when I hack something in at 2am!

**Comment by HeneryH** [comment]
I took a step back and tried to test out accelerated ffmpeg from the command line.  To start with the most simple case, I spun up a clean bare metal ubuntu server, installed jellfin (to get the benefit of jellyfin-ffmpeg) and then tried out some command line testing.

A couple of observations:

1. The path for jellyfin-ffmpeg is in a non-standard location (/usr/lib/jellyfin-ffmpeg/ffmpeg) so just installing it and relying on MediaCMS to magically find it is unrealistic.
2. Running ffmpeg from the command line and using intel acceleration fails due to permissions.  You need to either run it as root or add the "render" group to your user account.
3. Then command line acceleration works fine.

I think taking these lessons back to my ProxMox system where I spun an lxc container using the jellyfin scripts and then installing MediaCMS on top of that might have better luck.

I also saw a thread here that the maintainer is having time pressures and can't dedicate enough time to this project.  Maybe we can suggest some committable solutions for the helpers.py file even if they are commented blocks for intel-qsv and nvidia acceleration that the user can then uncomment if desired.

**Reply by HeneryH** [reply]
@dmurphynj Your command line works on Big Buck Bunny just fine and the intel acceleration is working

![Screenshot 2025-02-03 140911](https://github.com/user-attachments/assets/13a13bd0-5638-4d23-8d38-affa33d3e3e6)

and I took a crack at the helper but must have missed something.  I'll try to grab a diff of the helper.

**Reply by HeneryH** [reply]
Can you see the changes I made in this file?  

https://github.com/HeneryH/mediacms/commit/4f7441a75d0127028bf9afe8a046235f9e03cbe8

**Reply by HeneryH** [reply]
I'm used to old school programming where you could run a debugger and step through the code.  In this case, being unfamiliar with the programming language (python, I presume), I might try to use the printf() debugger method and try to determine the actual command being generated by the helper.py file.

**Reply by aphixe** [reply]
Any updates on this HeneryH I got intel quicksync gpu, right now i guess i am stuck with gpu renders.

**Comment by HeneryH** [comment]
No, I never got it to work from MediaCMS.  I can get accelerated processing
from the command line but not within the tool.  The code referenced here in
this thread builds a command line command to process it and I never got it
to build me a good command that works.

**Comment by mjbergman92** [comment]
There's probably an obvious answer for those more familiar with the code base, but I'm just curious, so take this with a grain of salt.

Why not use environment variables for the entirety of ffmpeg options, with the current options in the code as the default?
If that's too much, what is the state of the experimental work being done to allow external transcoding?

Without being too familiar with the project, I don't see why adopting gpac with it's host of configuration options for queueing, etc. on top of ffmpeg would introduce any limitations in comparison to the "difficulty" to configure those operations using ffmpeg directly in the MediaCMS code base. Again, I'm more curious than anything. I just see that the contributions for gpac are steady and I don't think it's going away. There's 2 long-time contributors still active along with 3-5 total active contributors, depending on your definition.

**Comment by HeneryH** [comment]
I spent some time as an amateur coder on this a while back and didn´t make much progress.  Revisited the concept since I am rehositting from install.sh to docker.  I am getting lazier and just letting AI fix things for me now...

> the stock ffmpeg in mediacms/mediacms:latest is a static build with no VAAPI/QSV/NVENC compiled in at all, so no amount of correct helpers.py code could ever have worked. That's almost certainly what defeated your own attempt and everyone else's in that thread.
> 
> Fix: Built a custom image layering jellyfin-ffmpeg (the build Jellyfin maintains specifically because it has full hardware-accel support compiled in) onto the base image, wired the Intel iGPU (now enabled in BIOS) into the container via device bind-mount + group_add, and patched files/helpers.py with a software-decode/hardware-encode-only VAAPI pipeline (the safest, most compatible pattern — avoids the segfaults others hit with full hardware pipelines).
